### PR TITLE
dial: limit error size

### PR DESCRIPTION
**Pull request by Stebalien**
Instead of storing _every_ error, store at most 32 errors (plus a "too many errors" error).

Helps address https://github.com/libp2p/go-libp2p-swarm/issues/119

Note: https://github.com/libp2p/go-libp2p-kad-dht/pull/328 actually _fixes_ the issue but this PR helps reduce the impact of similar issues.

### CODE REVIEW

File: swarm_dial.go [hunk 1]
```diff
@@ -58,6 +61,9 @@ const ConcurrentFdDials = 160
 // per peer
 const DefaultPerPeerRateLimit = 8
 
+// maxDialErrors is the maximum number of dial errors we record
+const maxDialErrors = 32
```
Comment (vyzo) [inline swarm_dial.go]: this might even be too many; maybe 16 or 10?

**Comment by raulk** [comment]
For usability, it would be cool to:
- include the value of `maxDialErrors` inside the message for `ErrTooManyErrors`, e.g. "too many errors (max tracked: 32)".
- turn `ErrTooManyErrors` into a struct that counts _how many more errors_ we found beyond our max; when we reach the threshold, we can retain a pointer to it and keep incrementing the counter for every additional error. That way we could at least tell the user how many more errors we saw but failed to retain.

**Comment by Stebalien** [comment]
Alternative here: https://github.com/libp2p/go-libp2p-swarm/pull/120.

**Comment by Stebalien** [comment]
(closing in favor of the alternative, we can reopen if necessary)

**Pull request closed without merging**